File: lib/gherkin/formatter/step_printer.rb
module Gherkin
  module Formatter
    class StepPrinter
      def write_step(io, step_format, step_name, arguments)
        unpacked_step_name = step_name.unpack("U*")
        
        text_start = 0
        arguments.each do |arg|
          text = unpacked_step_name[text_start..arg.offset-1].pack("U*")
          step_format.write_text(io, text) unless arg.offset == 0
          step_format.write_arg(io, arg.val)
          text_start = arg.offset + arg.val.unpack("U*").length
        end
        step_format.write_text(io, unpacked_step_name[text_start..-1].pack("U*")) unless text_start == unpacked_step_name.length
      end
    end
  end
end

File: java/src/main/java/gherkin/formatter/StepPrinter.java
package gherkin.formatter;

import java.util.List;

public class StepPrinter {
    public void writeStep(NiceAppendable out, Format textFormat, Format argFormat, String stepName, List<Argument> arguments) {
        int textStart = 0;
        for (Argument argument : arguments) {
            // can be null if the argument is missing.
            if (argument.getOffset() != null) {
                String text = stepName.substring(textStart, argument.getOffset());
                out.append(textFormat.text(text));
            }
            // val can be null if the argument isn't there, for example @And("(it )?has something")
            if (argument.getVal() != null) {
                out.append(argFormat.text(argument.getVal()));
                textStart = argument.getOffset() + argument.getVal().length();
            }
        }
        if (textStart != stepName.length()) {
            String text = stepName.substring(textStart, stepName.length());
            out.append(textFormat.text(text));
        }
    }
}
